### PR TITLE
sidecar: allow profiling memory use over time

### PR DESCRIPTION
**Pull request by stevekuznetsov**
For investigating the memory consumption of `sidecar` over time, it's
now possible to capture a heap profile every 30 seconds and plot the
results over time.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/assign @cjwagner @alvaroaleman 

Example plot:
Screenshot from 2021-03-19 11-32-59

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Interestingly, even though `sidecar` has absolutely nothing to do with k8s, somehow we side-effect import something that sets up a `Scheme` in some `init()` ...

File: hack/analyze-memory-profiles.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python3
```
Comment (alvaroaleman) [inline hack/analyze-memory-profiles.py:1]: Can you document what this does?

File: prow/sidecar/options.go [hunk 2]
```diff
@@ -76,6 +76,9 @@ type Options struct {
 	// taken by `sidecar` to upload all relevant artifacts.
 	IgnoreInterrupts bool `json:"ignore_interrupts,omitempty"`
 
+	// WriteMemoryProfile makes the program write a memory profile at process exit.
+	WriteMemoryProfile bool `json:"write_memory_profile,omitempty"`
```
Comment (alvaroaleman) [inline prow/sidecar/options.go:80]: Can you mention the script?

**Comment by alvaroaleman** [comment]
> Interestingly, even though sidecar has absolutely nothing to do with k8s, somehow we side-effect import something that sets up a Scheme in some init() ...

Does go allow to analyze why a given binary depends on a given package? I only know how to do that for the entire module, but that is useless here

**Comment by stevekuznetsov** [comment]
I do not know for sure about tracing @alvaroaleman .


... 

mmm my favorite thing, `pylint`!

**Comment by stevekuznetsov** [comment]
```
 Linting python...
************* Module hack.analyze-memory-profiles
hack/analyze-memory-profiles.py:13:0: E0401: Unable to import 'matplotlib.font_manager' (import-error)
hack/analyze-memory-profiles.py:14:0: E0401: Unable to import 'matplotlib.dates' (import-error)
hack/analyze-memory-profiles.py:15:0: E0401: Unable to import 'matplotlib.pyplot' (import-error)
hack/analyze-memory-profiles.py:16:0: E0401: Unable to import 'matplotlib.ticker' (import-error)
-----------------------------------
Your code has been rated at 9.93/10 
```

?????

File: hack/verify-pylint.sh [hunk 3]
```diff
@@ -37,4 +37,4 @@ export PYLINTHOME=$TEST_TMPDIR
 shopt -s extglob globstar
 
 # TODO(clarketm): remove `boskos` exclusion after upgrading to PY3.
-"$DIR/pylint_bin" !(gubernator|external|vendor|jenkins|scenarios|triage|boskos|bazel-*)/**/*.py
+"$DIR/pylint_bin" !(gubernator|external|vendor|jenkins|scenarios|triage|boskos|bazel-*|hack/analysis)/**/*.py
```
Comment (stevekuznetsov) [inline hack/verify-pylint.sh:40]: @BenTheElder this doesn't seem to have the desired effect :|

**Review by alvaroaleman — APPROVED** [review]

**Pull request merged**